File: kiosk/src/lib/rfid.test.ts
import { expect, it } from 'vitest';

import { padRfid, RFID_LENGTH } from './rfid';

it('pads a short numeric tag with leading zeros to 24 characters', () => {
  const { tag, problem } = padRfid('100348');
  expect(problem).toBeNull();
  expect(tag).toBe('000000000000000000100348');
  expect(tag).toHaveLength(RFID_LENGTH);
});

it('leaves an already-24-character tag alone, and is idempotent', () => {
  const stored = padRfid('100348').tag!;
  expect(padRfid(stored).tag).toBe(stored);
  expect(padRfid('E2004321' + '0'.repeat(16)).tag).toBe('E2004321' + '0'.repeat(16));
});

it('upper-cases and strips whitespace, inside as well as around', () => {
  expect(padRfid('  e200 4321 ').tag).toBe('0'.repeat(16) + 'E2004321');
});

it('reports an empty value rather than padding it to 24 zeros', () => {
  for (const raw of ['', '   ', '\t\n']) {
    expect(padRfid(raw)).toEqual({ tag: null, problem: 'empty' });
  }
});

it('rejects anything that is not alphanumeric', () => {
  for (const raw of ['E200-4321', '1003.48', 'tag_1', 'é200']) {
    expect(padRfid(raw)).toEqual({ tag: null, problem: 'not_alphanumeric' });
  }
});

it('rejects a tag longer than 24 characters', () => {
  expect(padRfid('1'.repeat(25))).toEqual({ tag: null, problem: 'too_long' });
  expect(padRfid('1'.repeat(24)).problem).toBeNull();
});

File: Android_Kiosk_App/app/src/test/java/com/serversherpa/kiosk/core/scan/RfidTest.kt
package com.serversherpa.kiosk.core.scan

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class RfidTest {
    @Test fun padsTo24AndUppercases() {
        val p = padRfid(" 10 03 48 ")
        assertEquals("000000000000000000100348", p.tag)
        assertNull(p.problem)
        assertEquals("00000000000000000000ABCD", padRfid("abcd").tag)
    }

    @Test fun problems() {
        assertEquals(RfidProblem.EMPTY, padRfid("   ").problem)
        assertEquals(RfidProblem.NOT_ALPHANUMERIC, padRfid("10-03").problem)
        assertEquals(RfidProblem.TOO_LONG, padRfid("1".repeat(25)).problem)
        assertEquals("Scan the RFID tag.", rfidProblemText(RfidProblem.EMPTY))
        assertEquals("That tag is longer than 24 characters.", rfidProblemText(RfidProblem.TOO_LONG))
    }

    @Test fun displayStripsLeadingZerosButKeepsOne() {
        assertEquals("100348", displayRfid("000000000000000000100348"))
        assertEquals("0", displayRfid("0000"))
        assertEquals("—", displayRfid(null))
        assertEquals("—", displayRfid(""))
    }
}
